R/add.R

```diff
@@ -137,16 +137,20 @@ add_column <- function(.data, ..., .before = NULL, .after = NULL) {
 
   pos <- pos_from_before_after_names(.before, .after, colnames(.data))
 
+  end_pos <- ncol(.data) + seq_len(ncol(df))
+
   if (pos <= 0L) {
-    out <- cbind(df, .data)
-  } else if (pos >= ncol(.data)) {
-    out <- cbind(.data, df)
+    indexes <- c(end_pos, seq_len(ncol(.data)))
+  } else if (pos > ncol(.data)) {
+    indexes <- seq_len(ncol(.data) + ncol(df))
   } else {
-    indexes <- seq_len(pos)
-    out <- cbind(.data[indexes], df, .data[-indexes])
+    indexes_before <- seq_len(pos)
```

Comment:
How about:

```r
indexes_before <- rlang::seq2(1L, pos)
indexes_after <- rlang::seq2(pos + 1L, ncol(.data))
```

This might just work in all three cases.

Reply:
Cleverer than I am. I think it makes sense to floor and cap the `pos` too just in case?

```r
  if(pos < 0L) {
    pos <- 0L
  } else if (pos > ncol(.data)) {
    pos <- ncol(.data)
  }

  indexes_before <- rlang::seq2(1L, pos)
  indexes_after <- rlang::seq2(pos + 1L, ncol(.data))
  indexes <- c(indexes_before, end_pos, indexes_after)
```

This would fix the following two problems that occur with the current soln:

```r
# Duplicated because indexes becomes c(1,2,3,4,5,6,6)
> iris %>% add_column(x = iris$Sepal.Width, .after = 6) %>% head
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species   x x.1
1          5.1         3.5          1.4         0.2  setosa 3.5 3.5
2          4.9         3.0          1.4         0.2  setosa 3.0 3.0
3          4.7         3.2          1.3         0.2  setosa 3.2 3.2
4          4.6         3.1          1.5         0.2  setosa 3.1 3.1
5          5.0         3.6          1.4         0.2  setosa 3.6 3.6
6          5.4         3.9          1.7         0.4  setosa 3.9 3.9
```

```r
> iris %>% add_column(x = iris$Sepal.Width, .after = -2) %>% head
Error in `[.default`(.data, indexes) : 
  only 0's may be mixed with negative subscripts

# This happens because `indexes_after` contains
[1] -1  0  1  2  3  4  5
```

Reply:
Yeah, I forgot that `seq2()` only takes care of one side of the capping. Using `min()` and `max()` could make the code a tad shorter, but both solutions work.